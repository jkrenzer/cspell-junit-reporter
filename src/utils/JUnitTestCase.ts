export class JUnitTestCase {
    name: string;
    classname: string;
    file: string;
    line: number;
    error_message: string;
    error_type: string;
    error_context: string;

    constructor(name: string, classname: string, file: string, line: number, error_message: string = "", error_type: string = "", error_context: string = "") {
        this.name = name;
        this.classname = classname;
        this.file = file;
        this.line = line;
        this.error_message = error_message;
        this.error_type = error_type;
        this.error_context = error_context;
    }

    public toXMLObj() {
        let testcase_attr = {
            name: this.name,
            classname: this.classname,
            file: this.file,
            line: this.line,
        }
        if (this.error_type) {
            return {
                testcase: {
                    $: testcase_attr,
                    failure: {
                        $: {
                            message: this.error_message,
                            type: this.error_type
                        },
                        _: this.error_context || ""

                    }
                }
            };
        }
        else {
            return {
                testcase: {
                    $: testcase_attr
                }
            };
        }
    }

};