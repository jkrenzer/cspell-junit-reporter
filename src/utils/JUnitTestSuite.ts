import { JUnitTestCase } from "./JUnitTestCase.js";

export class JUnitTestSuite {
    name: string;
    testcases: Array<JUnitTestCase>;


    constructor(name: string, testcases: Array<JUnitTestCase>) {
        this.name = name;
        this.testcases = testcases;
    }

    public toXMLObj() {
        let successes = 0;
        let failures = 0;
        let testcases = this.testcases.map(
            function (testcase) {
                if (testcase.error_type != "")
                    failures++;
                else
                    successes++;
                return testcase.toXMLObj().testcase; // Cut the top key to merge in root's key
            });
        let tests = failures + successes;
        return {
            testsuite: {
                $: {
                    name: this.name,
                    tests: tests,
                    failures: failures,
                    errors: 0,
                    skipped: 0,
                    assertions: tests,
                    time: 0.0,
                    file: "./",
                    timestamp: (new Date()).toISOString()

                },
                testcase: testcases

            }
        };
    }

};