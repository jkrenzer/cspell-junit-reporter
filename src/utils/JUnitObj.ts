import { JUnitTestSuite } from './JUnitTestSuite.js';

export class JUnitObj {
    testsuites: Array<JUnitTestSuite>;

    constructor(testsuites: Array<JUnitTestSuite>) {
        this.testsuites = testsuites;
    }

    public toXMLObj() {
        let tests = 0;
        let failures = 0;
        let testsuites = this.testsuites.map(
            function (testsuit) {
                let obj = testsuit.toXMLObj().testsuite; // Cut the top key to merge in root's key
                tests += obj.$.tests || 0;
                failures += obj.$.failures || 0;
                return obj;
            }
        );
        return {
            testsuites: {
                $: {
                    failures: failures,
                    tests: tests
                },
                testsuite: testsuites
            }
        }
    }
}