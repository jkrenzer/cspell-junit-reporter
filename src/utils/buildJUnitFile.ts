import { Builder } from "xml2js";

import {
    ErrorLike,
    Issue,
    IssueType,
    MessageType,
    ProgressFileComplete,
    ProgressItem,
    RunResult,
} from '@cspell/cspell-types';

import { JUnitTestCase } from "./JUnitTestCase.js";
import { JUnitTestSuite } from "./JUnitTestSuite.js";
import { JUnitObj } from "./JUnitObj.js";

function getIssueType(issue: Issue) {
    if (issue.issueType != undefined && issue.issueType == IssueType.spelling)
        return "spelling";
    else
        return "directive";
}

function getErrorText(issue: Issue) {
    let text = issue.text || "";
    if (issue.suggestions != undefined)
        return text + "\n\nSuggestions:\n" + issue.suggestions;
    else
        return text;
}

export function makeJUnitObj(issues: Array<Issue>) {
    let testcases: Array<JUnitTestCase> = [];
    if (Array.isArray(issues) && issues.length > 0) {
        issues.forEach(
            function (issue, idx) {
                testcases.push(
                    new JUnitTestCase(
                        "SpellingError" + idx,
                        "Spellcheck",
                        issue.uri || "Unknown",
                        issue.line.offset,
                        issue.message || "Spelling error detected",
                        getIssueType(issue),
                        getErrorText(issue)
                    )
                )
            }
        )
    }
    else {
        testcases.push(
            new JUnitTestCase(
                "Spelling",
                "Spellcheck",
                "Unknown",
                0
            )
        )
    }
    let testsuit: JUnitTestSuite = new JUnitTestSuite("CSpellChecks",
        "Spellchecks", testcases);
    return new JUnitObj([testsuit]);
}

export function toXML(obj: JUnitObj | JUnitTestCase | JUnitTestSuite) {
    let builder = new Builder({
        cdata: true
    });
    return builder.buildObject(obj.toXMLObj());
}

export function makeJUnitXML(issues: Array<Issue>) {
    return toXML(makeJUnitObj(issues));
}