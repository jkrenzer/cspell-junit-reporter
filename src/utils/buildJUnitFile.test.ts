import { describe, expect, test } from 'vitest';

import { JUnitObj } from "./JUnitObj.js";
import { JUnitTestSuite } from "./JUnitTestSuite.js";
import { JUnitTestCase } from "./JUnitTestCase.js";

import { toXML } from './buildJUnitFile.js';

describe('toXMLObj', () => {
    let testcase1 = new JUnitTestCase('name', 'classname', 'file', 0, 'error_message', 'error_type', 'error_context');
    let testcase2 = new JUnitTestCase('name', 'classname', 'file', 0);
    test.each([
        { Obj: new JUnitObj([]) },
        { Obj: new JUnitTestSuite('name', 'classname', []) },
        { Obj: testcase1 },
        { Obj: testcase2 }
    ])('$Obj minimal -> Obj', ({ Obj }) => {
        let obj = Obj.toXMLObj();
        expect(obj).toMatchSnapshot();
        expect(() => toXML(Obj)).not.toThrow();
        let xmlObj = toXML(Obj);
        expect(xmlObj).toMatchSnapshot();
    });
    let testsuite = new JUnitTestSuite('name', 'classname', [testcase1, testcase2]);
    let testobj = new JUnitObj([testsuite]);
    test.each([
        { Obj: testobj },
        { Obj: testsuite },
        { Obj: testcase1 },
        { Obj: testcase2 }
    ])('$Obj inkremental -> Obj', ({ Obj }) => {
        let obj = Obj.toXMLObj();
        expect(obj).toMatchSnapshot();
        expect(() => toXML(Obj)).not.toThrow();
        let xmlObj = toXML(Obj);
        expect(xmlObj).toMatchSnapshot();
    });
    test.each([
        { Obj: testsuite, root: "testsuite" },
        { Obj: testobj, root: "testsuites" }
    ])('Failure count calculation', ({ Obj, root }) => {
        let obj = Obj.toXMLObj();
        let attr: object = obj[root].$;
        expect(attr.failures).toEqual(1);
        expect(attr.tests).toEqual(2);
    });

});