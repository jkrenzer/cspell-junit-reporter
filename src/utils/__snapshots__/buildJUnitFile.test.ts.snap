// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toXMLObj > JUnitObj{ testsuites: [ …(1) ] } inkremental -> Obj 1`] = `
{
  "testsuites": {
    "$": {
      "failures": 1,
      "tests": 2,
    },
    "testsuite": [
      {
        "$": {
          "assertions": 2,
          "errors": 0,
          "failures": 1,
          "file": "./",
          "name": "name",
          "skipped": 0,
          "tests": 2,
          "time": 0,
          "timestamp": "2024-04-22T17:33:42.238Z",
        },
        "testcase": [
          {
            "$": {
              "assertions": 1,
              "classname": "classname",
              "file": "file",
              "line": 0,
              "name": "name",
              "time": 0,
            },
            "failure": {
              "$": {
                "message": "error_message",
                "type": "error_type",
              },
              "_": "error_context",
            },
          },
          {
            "$": {
              "assertions": 1,
              "classname": "classname",
              "file": "file",
              "line": 0,
              "name": "name",
              "time": 0,
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`toXMLObj > JUnitObj{ testsuites: [ …(1) ] } inkremental -> Obj 2`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testsuites failures="1" tests="2">
  <testsuite name="name" tests="2" failures="1" errors="0" skipped="0" assertions="2" time="0" file="./" timestamp="2024-04-22T17:33:42.239Z">
    <testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0">
      <failure message="error_message" type="error_type">error_context</failure>
    </testcase>
    <testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0"/>
  </testsuite>
</testsuites>"
`;

exports[`toXMLObj > JUnitObj{ testsuites: [] } minimal -> Obj 1`] = `
{
  "testsuites": {
    "$": {
      "failures": 0,
      "tests": 0,
    },
    "testsuite": [],
  },
}
`;

exports[`toXMLObj > JUnitObj{ testsuites: [] } minimal -> Obj 2`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testsuites failures="0" tests="0"/>"
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } inkremental -> Obj 1`] = `
{
  "testcase": {
    "$": {
      "assertions": 1,
      "classname": "classname",
      "file": "file",
      "line": 0,
      "name": "name",
      "time": 0,
    },
    "failure": {
      "$": {
        "message": "error_message",
        "type": "error_type",
      },
      "_": "error_context",
    },
  },
}
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } inkremental -> Obj 2`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0">
  <failure message="error_message" type="error_type">error_context</failure>
</testcase>"
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } inkremental -> Obj 3`] = `
{
  "testcase": {
    "$": {
      "assertions": 1,
      "classname": "classname",
      "file": "file",
      "line": 0,
      "name": "name",
      "time": 0,
    },
  },
}
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } inkremental -> Obj 4`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0"/>"
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } minimal -> Obj 1`] = `
{
  "testcase": {
    "$": {
      "assertions": 1,
      "classname": "classname",
      "file": "file",
      "line": 0,
      "name": "name",
      "time": 0,
    },
    "failure": {
      "$": {
        "message": "error_message",
        "type": "error_type",
      },
      "_": "error_context",
    },
  },
}
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } minimal -> Obj 2`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0">
  <failure message="error_message" type="error_type">error_context</failure>
</testcase>"
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } minimal -> Obj 3`] = `
{
  "testcase": {
    "$": {
      "assertions": 1,
      "classname": "classname",
      "file": "file",
      "line": 0,
      "name": "name",
      "time": 0,
    },
  },
}
`;

exports[`toXMLObj > JUnitTestCase{ name: 'name', …(6) } minimal -> Obj 4`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0"/>"
`;

exports[`toXMLObj > JUnitTestSuite{ name: 'name', …(1) } inkremental -> Obj 1`] = `
{
  "testsuite": {
    "$": {
      "assertions": 2,
      "errors": 0,
      "failures": 1,
      "file": "./",
      "name": "name",
      "skipped": 0,
      "tests": 2,
      "time": 0,
      "timestamp": "2024-04-22T17:33:42.240Z",
    },
    "testcase": [
      {
        "$": {
          "assertions": 1,
          "classname": "classname",
          "file": "file",
          "line": 0,
          "name": "name",
          "time": 0,
        },
        "failure": {
          "$": {
            "message": "error_message",
            "type": "error_type",
          },
          "_": "error_context",
        },
      },
      {
        "$": {
          "assertions": 1,
          "classname": "classname",
          "file": "file",
          "line": 0,
          "name": "name",
          "time": 0,
        },
      },
    ],
  },
}
`;

exports[`toXMLObj > JUnitTestSuite{ name: 'name', …(1) } inkremental -> Obj 2`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testsuite name="name" tests="2" failures="1" errors="0" skipped="0" assertions="2" time="0" file="./" timestamp="2024-04-22T17:33:42.242Z">
  <testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0">
    <failure message="error_message" type="error_type">error_context</failure>
  </testcase>
  <testcase name="name" classname="classname" file="file" line="0" assertions="1" time="0"/>
</testsuite>"
`;

exports[`toXMLObj > JUnitTestSuite{ name: 'name', …(1) } minimal -> Obj 1`] = `
{
  "testsuite": {
    "$": {
      "assertions": 0,
      "errors": 0,
      "failures": 0,
      "file": "./",
      "name": "name",
      "skipped": 0,
      "tests": 0,
      "time": 0,
      "timestamp": "2024-04-22T17:33:42.233Z",
    },
    "testcase": [],
  },
}
`;

exports[`toXMLObj > JUnitTestSuite{ name: 'name', …(1) } minimal -> Obj 2`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<testsuite name="name" tests="0" failures="0" errors="0" skipped="0" assertions="0" time="0" file="./" timestamp="2024-04-22T17:33:42.235Z"/>"
`;
